### DISCUSSION TITLE
Convex Least Squares as the backbone of AA

### DISCUSSION THREAD

**Discussion by wangzcl** [aleixalcacer/archetypes#24] [General]
In multiple variants of AA, the core core optimization process involves alternatively solving least squares under simplex constraints, or convex least squares (CLS):

mininize $||Ax-b||_F^2$

subject to $x_i\geq 0$ and $\sum_i x_i=1$

An approximate algorithm is a penalized non-negative least square proposed by Culter and Breiman (1994). A list of some exact algorithms can be found here: https://lcondat.github.io/publis/Condat-unmixing.pdf.

In some fields (e.g., geochemistry), we also use CLS directly when the archetypes of the dataset are already known. However, in some applications, people use only a loose version of CLS by simply adding $\sum_i x_i=1$ into the equation set.

Perhaps we can make CLS the backbone of AA development: First, implementing different algorithms of CLS, e.g. ``_cls_penalized_nnls``, ``_cls_pgd``, ``_cls_jax`` etc and also provide a ``cls(A, b, method=...)`` function at user level. We can then implement AA by calling ``cls`` to optimize $A$ and $B$ alternatively and specifying the CLS algorithm by passing the ``method`` parameter to AA.

This may simplify the structure of the source code, and make it more flexible when we try to implement more variants of AA. For example: An optimization of original AA is proposed by Bauckhage and Thurau (2009), in which we only consider points on the boundary of dataset's convex hull as potential archetypes. Here, The way to optimize $A$ is the same as original AA, while optimization of $B$ is slightly different, but is still within the framework of CLS. Using CLS as the backbone will free us from rewriting all the ``AAOptimizer``s when implementing new AA variants.

However, this practice can possibly lead to performance sacrifice, as sometimes we have to manually transpose arrays before inputting them into ``cls``. I think this can be alleviated if we take of advantage of existing optimization librarys, compile code in Cython and improve the code at algorithm level.

**Comment by aleixalcacer** [comment]
I like the idea of implementing different cls and provide them in the package. Regarding your problem, it can be solved by passing the archetypes in the initialization and then call directly to transform (without fit). To do that, we could adapt the init parameter in the constructor to handle a matrix of user-defined archetypes as initialization. What do you think?

But kudos to implement multiple cls and use it inside the algorithms :)

**Reply by wangzcl** [reply]
This makes a lot of sense. It would be great if we could allow users to manually initialize archetypes, both for the task I mentioned and for general purpose. The main reason for me to propose using ``cls`` is that, I think it can potentially simplify the structure of the code (maybe even allow us to skip the ``AAOptimizer`` level, not sure though)

**Comment by wangzcl** [comment]
I have an idea about projected gradient method. PGD requires orthogonal projection onto a probability simplex. It looks like currently we only have
```
C_ = np.where(C_ < 1e-8, 0, C_)
C_ = C_ / np.sum(C_, axis=0)
```
There may be possible issues here. https://math.stackexchange.com/questions/2005154/how-to-project-onto-the-unit-simplex-as-intersection-of-two-sets-optimizing-a-c Someone has already discussed the exactly same topic. 
Projection onto a probability simplex seems to be a non-trivial problem. The best solution I can find is Condat 2015: https://link.springer.com/article/10.1007/s10107-015-0946-6. If we continue with the idea of implementing ``cls``, I can try translating Condat's code into ``Cython`` and incorporate it into the package. After that we can do PGD with this. Though there seems to be more effecient algorithm, I think we are fine with PGD.
What do you think? @aleixalcacer

**Reply by aleixalcacer** [reply]
The fix introduced in 5d85dc49aed05cf5dcb86b110cb3b99b0f88ddff should solve it. However, if you want to implement the 5 algorithms that are discussed in Condat 2015, I will be thankful. We could add a new parameter in the AA initialization to let the user choose the one they want. Do you agree?

**Reply by wangzcl** [reply]
As what is mentioned in the stackexchange post, the composition of projections is not the projection onto the intersection of sets. I think that may be the issue here. However, if the current implementation proves to work fine, I am ok with it.
I will do a little test to see if the current method works.

**Reply by wangzcl** [reply]
Here is what I mean:
1. PGD requires orthogonal projection to the feasible set (a unit/probability simplex in our case). Example:
    We search for a minimizer of $f(x,y)$ on a plane in a feasible set (blue triangle). Let's say the minimizer is point $D$, on the boundary of the feasible set. The negative gradient $\boldsymbol{u} = -\nabla f |_D$ must be orthogonal to the boundary of the feasible set, otherwise there will be a non-zero projection of $\boldsymbol{u}$ along the boundary of the feasible set, which means that $D$ is not a minimizer.
    If PGD is correctly performed, it should converge at $D$, which means any further PGD steps will bring us back to $D$. A unconstrained projected descend will take us to $E$. And then an orthogonal projection onto the feasible set will take us back to $D$. However, if the projection is non-orthogonal, for example, the projection of $E$ is $F$, it will take us away from the minimizer $D$. The algorithm may still converge, but not at the real minimizer.
![ortho_proj](https://github.com/aleixalcacer/archetypes/assets/43379393/1348beda-7fb6-43a2-b99f-c0d3550a036c)

2. $ Projection onto $\mathbb{R}^n_{+}$ plus a rescaling by sum is not an orthogonal projection onto unit simplex. Example:
    We want to project $C$ and $E$ onto 2D unit simplex $\bar{AB}$. The right answer should be $F$ and $I$, respectively. If we simply make the negative values zero and then divide by sum, we will get $C\to D \to A$ and $E \to H$.
![geogebra-export (1)](https://github.com/aleixalcacer/archetypes/assets/43379393/a4e39f96-b2bc-4035-aebf-10ec72f51e6b)

If I made any mistakes, please do let me know.

**Reply by aleixalcacer** [reply]
I'm not contradicting you at all. On the contrary, I agree with what you're saying.

The reason I want to keep the alternating method (first non-negative, then convex) is because it's used in Morten Morup's paper. However, I am in favor of implementing another Projected Gradient Descent that performs direct projection, as you suggest. Do you think that's OK?

**Reply by wangzcl** [reply]
Thanks for clarifying that. As the alternating projection proved to work fine in Morup paper, we can definitely keep it as an option.

**Reply by wangzcl** [reply]
I took a look at Morup's PGD. It seems that though the projection is not orthogonal, they fixed the problem I showed in the figure by adding another term in the gradient. Besides, it seems that they don't use $A^+$ in the optimization of $B$. Instead they use the explicit form of $\nabla_B ||X-ABX||$, which makes more sense (not sure if I understood it correctly).
The difference way of optimizing $B$ makes these AA variants totally different. Now I no longer expect that ``cls`` can unify the implementations of different AA variants. However, I think it's still make sense to implement different ``cls``. Now we have different versions of ``cls``: 1. penalized NNLS; 2. Condat PGD; 3. Morup PGD. Perhaps we can let the users choose which one to use in the Eugster&Leisch version of AA, as the optimization of $B$ there can be simply translated to ``cls``. This may potentially increase the performance of the algorithm and/or makes it more numerically stable. As for Morup AA, I do agree that there should be a separate function to optimize $B$.
The question is, do we need to implement AA variants as subclasses of ``AABase`` instead of letting users to choose optimizers? For a certain AA architecture, e.g., Eugster&Leisch, there can be different ``cls`` and thus different optimizing methods. And then there can be a top-level class ``AA`` to wrap all these variants or the best one.

**Comment by aleixalcacer** [comment]
On another note, I'm going to merge my dev branch into main. I think it's ready. I'm still not 100% satisfied with the API, but it's better than nothing.

Additionally, I'm not entirely convinced that making `cls` the base of the package is the best in terms of performance, since for the `pgd` and `jax`, the algorithms differ depending on whether we are calculating A or B because they are optimized differently.

**Reply by wangzcl** [reply]
I agree that using that using ``cls`` as the base may sacrify some performance. I guess on of the main reasons for that would be matrix transportation and matrix continuity. There is a trade-off between how much performance we lose and how the structure, flexibility and extendability of the code can be improved. If we only lose, for example, 20% of the performance, I think it is acceptable. This need to be checked though. Besides, we have other ways to improve the performance. I will be happy to convert some low-level loops to ``cython``.

Another question that popped up in my mind is that ``jax``, like ``pytorch``, uses a different array data structure. Why it is integrated into the ``sklearn`` api instead of having its own folder? If a ``sklearn`` model takes `numpy` arrays as input, I suppose that a ``jax`` optimizer may perform conversion in the background. Is that true?

**Reply by aleixalcacer** [reply]
I hadn't thought about it. From what I see, both Jax and PyTorch implement the array protocol, that is, they avoid copying the data when converting between objects. so I'm thinking about fixing the sklearn api, and then implementing the different backends. I don't know whether to implement everything with the different backends or simply use PyTorch and Jax for the autograd. what do you think?

**Reply by wangzcl** [reply]
I think it really depends on our goal. If you are planning to merge AA into sklearn in the future, perhaps they won't want the module to depend on Jax or Pytorch. Otherwise it would be OK to use these packages in optimization and then explicitly convert the result into NumPy arrays.

**Reply by aleixalcacer** [reply]
I prefer to keep it as a separate package. What I don't know whether to do
```
from archetypes.jax import AA

model = AA(...)
```

or

```
from archetypes import AA

model = AA(..., backend="jax")
```

Maybe the second one?

**Reply by wangzcl** [reply]
I think the first one would be better. I presume that one of the most common cases for our users to use `jax` or `pytorch` is that they are, or they want to work with, GPU arrays. In that case, array conversion can be a problem. Besides, I am developing orthogonal projection onto simplex and corresponding PGD, and the projection function may not work with `jax` or `pytorch` arrays without array conversion. It would be great if we can separate sklearn, jax and pytorch so that I can focus on the sklearn and numpy part for now ...
